src/compiler/checker.ts

```diff
@@ -45578,6 +45637,297 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                /*withAlias*/ false,
+                node,
+                Diagnostics.The_return_type_of_an_async_function_must_either_be_a_valid_promise_or_must_not_contain_a_callable_then_member,
+            )
+            : narrowedExprType;
+        checkTypeAssignableToAndOptionallyElaborate(narrowedUnwrappedExprType, narrowedReturnType, errorNode, effectiveExpr);
+    }
+
+    function checkConditionalReturnExpression(
+        container: SignatureDeclaration,
+        returnType: Type,
+        node: ReturnStatement | Expression,
+        expr: ConditionalExpression,
+    ): void {
+        checkReturnExpression(container, returnType, node, expr.whenTrue, checkExpression(expr.whenTrue), /*inConditionalExpression*/ true);
+        checkReturnExpression(container, returnType, node, expr.whenFalse, checkExpression(expr.whenFalse), /*inConditionalExpression*/ true);
+    }
+
+    // Narrowable type parameters are type parameters that:
+    // (1) have a union type constraint;
+    // (2) are used as the type of a single parameter in the function, and nothing else
+    function getNarrowableTypeParameters(candidates: TypeParameter[]): [TypeParameter, Symbol, Identifier][] {
+        const narrowableParams: [TypeParameter, Symbol, Identifier][] = [];
+        for (const typeParam of candidates) {
+            const constraint = getConstraintOfTypeParameter(typeParam);
+            if (!constraint || !(constraint.flags & TypeFlags.Union)) continue;
+            if (typeParam.symbol && typeParam.symbol.declarations && typeParam.symbol.declarations.length === 1) {
+                const declaration = typeParam.symbol.declarations[0];
+                const container = isJSDocTemplateTag(declaration.parent) ? getJSDocHost(declaration.parent) : declaration.parent;
+                if (!isFunctionLike(container)) continue;
+                let reference: Identifier | undefined;
+                let hasInvalidReference = false;
+                for (const paramDecl of container.parameters) {
+                    const typeNode = getEffectiveTypeAnnotationNode(paramDecl);
+                    if (!typeNode) continue;
+                    if (isTypeParameterReferenced(typeParam, typeNode)) {
+                        let candidateReference;
+                        if (
+                            isTypeReferenceNode(typeNode) &&
+                            isReferenceToTypeParameter(typeParam, typeNode) &&
+                            (candidateReference = getValidParameterReference(paramDecl, constraint))
+                        ) {
+                            if (reference) {
+                                hasInvalidReference = true;
+                                break;
+                            }
+                            reference = candidateReference;
+                            continue;
+                        }
+                        hasInvalidReference = true;
+                        break;

```

Comment:
I refactored this a bit but kept the `break`s. They are an optimization more than anything, because as soon as we know the type parameter has an invalid reference, we can stop looking through the parameters. But I could get rid of that optimization if the code is still too confusing.